### ISSUE TITLE
When using rcedit on 7-Zip created archive, file gets corrupted

### ISSUE THREAD

**Issue by JonathanMATechServices** [electron/rcedit#69]
I am trying to change the resources of an sfx exe archive created with 7-Zip. When I do so however, the content of the archive is **_replaced_** with the files for the resources. Yes I said _replaced_. If I do this with a large archive, the file size of the file is reduced dramatically. See below for details.

1. With [7-Zip](http://www.7-zip.org/) installed, create an sfx archive with the GUI or the following command.
`"C:\Program Files\7-Zip\7z.exe" a -sfx7z.sfx "<path of archive to create ending in .exe>" "<path of file or folder to archive>"`
2. Use the 7-Zip GUI to open the archive and view its contents which should be as expected. Then close the GUI.
3. Run the following command.
`rcedit-x64.exe "<path of archive>" --set-version-string "LegalCopyright" "Copyright (c) 2017 <name>"`
4. Again use the 7-Zip GUI to open the archive. Where is the original file or folder that was archived?

**Comment by thomasmfields** [comment]
I'm also seeing issues with rcedit corrupting my executable after setting the version string. It could be this issue. Is it possible to get a fix?

**Comment by Mgamerz** [comment]
Same exact issue.

**Comment by romjacket** [comment]
yup.  executables created with 7zsd.sfx will be corrupted randomly.

**Comment by Draykee** [comment]
I use 7sfx files from: https://github.com/chrislake/7zsfxmm and the following approach:

1. `rcedit.exe 7zsd.sfx <options>`
2. `copy /b 7zsd.sfx + sfx_config.txt + MyPackage.7z MyPackage.exe`

You'r changing the icon of the SFX module which is in front of the archive.

**Comment by romjacket** [comment]
> You'r changing the icon of the SFX module which is in front of the archive.

yup.  fixed. thanks.

**Comment by JonathanMATechServices** [comment]
Oh, I get it! I was doing the commands Draykee posted in the opposite order and it was not working.

So maybe rcedit could be made to work even when the commands are completed the other way around so that other people would not have this problem?

**Comment by si458** [comment]
im having a similar issue with https://github.com/leafac/caxa
when i build an exe then try using rcedit to say add an icon,
its just `Fatal error: Unable to commit changes`
then when you run the command again,
it does apply the icon but the exe size has shrunk by 90% ???

**Comment by rversteegen** [comment]
Since rcedit relies on the `BeginUpdateResource`/`UpdateResource`/`EndUpdateResource` winapi functions to do all the actual work of modifying an exe, it appears those functions trim off anything past the "official" end of the .exe file (as defined by the file size in the header), which is a popular place to put extra program data, which will be ignored by the OS. So appears that the fix for this bug (and very likely also bug #91) would be to copy any such data out of the .exe and then paste it back on after calling `EndUpdateResource`.

**Comment by caesay** [comment]
> it appears those functions trim off anything past the "official" end of the .exe file 

Indeed that's correct. dotnet5/6 binaries also store the application code (as IL) after the "official end" of the native binary, but it's worth noting that, at least in the dotnet case, simply copying the data away beforehand and putting it back at the end afterwards will not help - you'll still end up with a corrupt exe. 

The "offset" pointing to this extra code stored at the end is written within the real executable code, so unless you know how to update this offset the exe will still be garbage. 

I don't know for sure that other approaches will suffer the same problem (eg. 7z) but I suspect so, and it will likely be impossible to find a universal solution that works for every exe using this approach.

**Comment by rversteegen** [comment]
Oh, yes, if there's an offset somewhere to the appended data then rcedit can't hope to support it. But that's why appending zip files to an exe is so popular, it works because the header is at the end of the file without needing to hardcode the offset. And so I expect that what I described will work in quite a lot of cases. Don't know about 7zip.

**Comment by fatherofinvention** [comment]
> I use 7sfx files from: https://github.com/chrislake/7zsfxmm and the following approach:
> 
> 1. `rcedit.exe 7zsd.sfx <options>`
> 2. `copy /b 7zsd.sfx + sfx_config.txt + MyPackage.7z MyPackage.exe`
> 
> You'r changing the icon of the SFX module which is in front of the archive.

Legend

**Comment by tpimh** [comment]
Also trying to modify 7z SFX, and rcedit never finishes execution stuck in some kind of infinite loop.

I am using 7z SFX from this repo: https://github.com/garyzyg/7zs-windows

Executing like this: `.\rcedit-x64.exe .\7zS.sfx --set-icon --set-requested-execution-level 'asInvoker'`
